cwl/src/main/scala/cwl/preprocessor/CwlPreProcessor.scala

```diff
@@ -295,27 +301,52 @@ class CwlPreProcessor(saladFunction: BFile => Parse[String] = saladCwlFile) {
                           unProcessedReferences: UnProcessedReferences,
                           processedReferences: ProcessedReferences,
                           breadCrumbs: Set[CwlReference]): Parse[ProcessedJsonAndDependencies] = {
-    val foldFunction = processCwlRunReference(unProcessedReferences, breadCrumbs) _
-
-    // TODO: it would be nice to accumulate failures here somehow (while still folding and be able to re-use
-    // successfully processed references, so I don't know if ErrorOr would work)
-    findRunReferences(saladedJson.json).foldLeft(processedReferences.validParse)(foldFunction) map { newKnownReferences =>
+    import cats.syntax.apply._
 
+    // Take the processed runs and inject them in the json
+    def inlineProcessedJsons(newKnownReferences: ProcessedReferences, inlinedRunWorkflows: Map[String, ProcessedJsonAndDependencies]) = {
       // Provide a function to swap the run reference with its json content
-      val lookupFunction = {
+      val lookupFunction: Json => Json = {
         json: Json => {
-          val fromMap = for {
+          val fromRunReferenceMap = for {
             asString <- json.asString
             reference <- asString.asReference
             embbeddedJson <- newKnownReferences.get(reference)
           } yield embbeddedJson.json
 
-          fromMap.getOrElse(json)
+          val fromInlinedWorkflow = for {
+            asObject <- json.asObject
+            id <- asObject.kleisli("id")
+            idAsString <- id.asString
+            embbeddedJson <- inlinedRunWorkflows.get(idAsString)
+          } yield embbeddedJson.processedJson
+
+          fromRunReferenceMap.orElse(fromInlinedWorkflow).getOrElse(json)
         }
       }
 
-      ProcessedJsonAndDependencies(root.steps.each.run.json.modify(lookupFunction)(saladedJson.json), newKnownReferences)
+      val flattenedJson = root.steps.each.run.json.modify(lookupFunction)(saladedJson.json)
+      
+      ProcessedJsonAndDependencies(flattenedJson, newKnownReferences ++ inlinedRunWorkflows.values.flatMap(_.processedDependencies))
```

Comment:
Just a heads up that #3567 updates preprocessing also. Your PR curretly looks like it's on track to be merged first-- though I am already rebased 😉

Send me a PM if you want me to wait for you to merge first.

Reply:
I'll race you, whoever gets green first 😄